waagent

```diff
@@ -2087,6 +2098,24 @@ class FreeBSDDistro(AbstractDistro):
         SetFileContents(location+"/ovf-env.xml", ovfxml)
         return retcode,out
 
+    def fallocate(self,filename,bytes):
+        """
+        Create a non-sparse file of that size. Overwrite existing file. BSD-compatible dd version.
+        
+        To make this a bit faster, a block size of 16M is used. After that, we use a single block to deal with the rest.
+        It might be worth discussing if this should be put here as a method -- probably some `if IsLinux():` is enough.
+        """
+        blocksize = 16 * 1024 * 1024
+        if type(bytes) is str:
+           bytes = int(bytes)
+
+        # We want to make the return value look like real.
+        exitcode = Run("dd if=/dev/zero of=" + filename + " bs=" + str(blocksize) + " count=" + str(bytes/blocksize))
+        if bytes%blocksize:
+            # Finish up using a single block.
+            exitcode = Run("dd if=/dev/zero of=" + filename + " bs=" + str(bytes%blocksize) + " count=1")
```

Comment:
Yes.. and the only clean `oflag` stuff seems like a GNUism, defeating the (potential but hard-to-find) point of providing FreeBSD compatibility. I don't really like the rounding up idea, since it messes up with the original design of providing kb-level options.

Besides `oflag`, another candidate in dd,`seek`, looks tricky too unless we fall back to the original `bs=1024` (or the dd default and common 'logical sector' size `bs=512`) design for filling the last blocks. Doing some `>>` operations in the shell command can have our appending job done, but we are not actually using `dd`'s precise blockfile processing abilities on our output then.

And still that question -- should I really write a method like this? To align with the `dd` behavior I added some redundant checks (at least in our original context) for file existence.

(And a correction: 'simply overwrite the first 16MB [or less] of the file with more zeroes, [truncates the file to that size,] and not achieve your goal).